cc2538_rf: implement Radio HAL

jia200x opened the pull request:
<!--
The RIOT community cares a lot about code quality.
Therefore, before describing what your contribution is about, we would like
you to make sure that your modifications are compliant with the RIOT
coding conventions, see https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions.
-->

### Contribution description
This PR adds an `ieee802154_hal` implementation for the CC2538 radios. It's based on the work of #14655. It also includes the definition of an `ieee802154_radio_hal` module to be able to switch between `netdev` and `ieee802154_hal` implementations (they are mutually exclusive).

The test application was adapted from #14655 for testing this radio. When more radios are there, the test application will be expanded.

<!--
Put here the description of your contribution:
- describe which part(s) of RIOT is (are) involved
- if it's a bug fix, describe the bug that it solves and how it is solved
- you can also give more information to reviewers about how to test your changes
-->


### Testing procedure

Run `tests/ieee802154_hal` test. Ideally this should be tested between 2 CC2538 radios. If that's not possible, it should be enough to use an AT86RF2XX or NRF52840 with #14655.

- Test setting channel and TX power with the `config_phy` command:
```
2020-08-12 18:24:49,390 #  config_phy
2020-08-12 18:24:49,393 # Usage: config_phy <channel> <tx_pow>
```

An out of range value for TX power should give an error

- Send data between nodes using the `txtsnd` command.
```
Usage: txtsnd <long_addr> <len>
```
It's possible to get the long address with the `print_addr` command.

- Perform CCa with the `cca` command:
```
2020-08-12 18:28:57,628 #  cca
2020-08-12 18:28:57,629 # CLEAR
> 2020-08-12 18:28:57,827 #  cca
2020-08-12 18:28:57,828 # CLEAR
> 2020-08-12 18:28:58,157 #  cca
2020-08-12 18:28:58,158 # CLEAR
> 2020-08-12 18:28:58,334 #  cca
2020-08-12 18:28:58,335 # BUSY
> 2020-08-12 18:28:58,473 #  cca
2020-08-12 18:28:58,474 # CLEAR
```

- Change the CCA mode and the threshold with the `config_cca` command:
```
config_cca <ed|cs|or|and> [ED threshold]>
```

Note the ED is in dBm. The ED threshold doesn't affect the Carrier Sense mode (CCA mode 2).

- Try different TX modes (CSMA-CA, CCA, direct transmission) using the `tx_mode` command:
```
Usage: tx_mode <csma_ca|cca|direct>
```

- Try different RX states (ACK on, ACK disabled, ACK with Frame Pending or Promiscuous mode) with the `rx_mode` command:
```
Usage: rx_mode <on|off|pend|promisc>
``` 

Use a sniffer to check if the receiver sends ACK. As expected, a radio with Promiscuous Mode will receive all kind of packets with correct FCS.

- Use the `caps` command to get the available caps for a given device:
```
2020-08-19 14:24:46,502 #  caps
2020-08-19 14:24:46,503 #                CAPS              
2020-08-19 14:24:46,505 # =================================
2020-08-19 14:24:46,507 # - Frame Retransmissions:        n
2020-08-19 14:24:46,509 #     * Info retries:             n
2020-08-19 14:24:46,511 # - Auto CSMA-CA:                 y
2020-08-19 14:24:46,513 # - 2.4 GHz band:                 y
2020-08-19 14:24:46,515 # - SubGHz band:                  n
2020-08-19 14:24:46,516 # - TX DONE indication:           y
2020-08-19 14:24:46,518 #     * ACK Timeout indication:   n
2020-08-19 14:24:46,519 # - RX_START indication:          n
2020-08-19 14:24:46,521 # - CCA Done indication:          y
```
<!--
Details steps to test your contribution:
- which test/example to compile for which board and is there a 'test' command
- how to know that it was not working/available in master
- the expected success test output
-->

- Use the `test_states` command to test state transitions and measure transition time. The important thing here is the TX-RX turnaround time (it should be equal or less than 12 symbols, 192 us, in order to comply with LIFS frames like the ACK):
```
2020-08-19 14:09:52,563 # test_states
2020-08-19 14:09:52,564 # Setting state to: TRX_OFF
2020-08-19 14:09:52,565 # 	Transition took 7 usecs
2020-08-19 14:09:52,566 # TRX_OFF-> RX
2020-08-19 14:09:52,578 # 	Transition took 194 usecs
2020-08-19 14:09:52,578 # RX-> TRX_OFF
2020-08-19 14:09:52,579 # 	Transition took 6 usecs
2020-08-19 14:09:52,580 # TRX_OFF-> TX
2020-08-19 14:09:52,581 # 	Transition took 7 usecs
2020-08-19 14:09:52,583 # TX-> TRX_OFF
2020-08-19 14:09:52,584 # 	Transition took 6 usecs
2020-08-19 14:09:52,584 # 
2020-08-19 14:09:52,585 # Setting state to: RX
2020-08-19 14:09:52,586 # 	Transition took 193 usecs
2020-08-19 14:09:52,587 # RX-> TRX_OFF
2020-08-19 14:09:52,588 # 	Transition took 7 usecs
2020-08-19 14:09:52,588 # TRX_OFF-> RX
2020-08-19 14:09:52,591 # 	Transition took 193 usecs
2020-08-19 14:09:52,591 # RX-> TX
2020-08-19 14:09:52,593 # 	Transition took 6 usecs
2020-08-19 14:09:52,593 # TX-> RX
2020-08-19 14:09:52,605 # 	Transition took 194 usecs
2020-08-19 14:09:52,606 # 
2020-08-19 14:09:52,607 # Setting state to: TX
2020-08-19 14:09:52,608 # 	Transition took 7 usecs
2020-08-19 14:09:52,609 # TX-> TRX_OFF
2020-08-19 14:09:52,610 # 	Transition took 6 usecs
2020-08-19 14:09:52,610 # TRX_OFF-> TX
2020-08-19 14:09:52,611 # 	Transition took 7 usecs
2020-08-19 14:09:52,612 # TX-> RX
2020-08-19 14:09:52,613 # 	Transition took 193 usecs
2020-08-19 14:09:52,613 # RX-> TX
2020-08-19 14:09:52,614 # 	Transition took 6 usecs
2020-08-19 14:09:52,614 # 
2020-08-19 14:09:52,616 # 	Transition took 193 usecs
```
Note this radio takes ~193 us to switch between RX-TX if the radio is not programmed to automatically go from TX to RX. However, since the symbol time is 16 us, this is not a problem for detecting the IEEE 802.15.4 preamble.

#14787 proves this :)

### Issues/PRs references
#13943 and friends
<!--
Examples: Fixes #1234. See also #5678. Depends on PR #9876.

Please use keywords (e.g., fixes, resolve) with the links to the issues you
resolved, this way they will be automatically closed when your pull request
is merged. See https://help.github.com/articles/closing-issues-using-keywords/.
-->

benpicco commented:
I wanted to test `examples/gnrc_networking` with the new HAL, so I added

```
USEMODULE += ieee802154
USEMODULE += ieee802154_radio_hal
USEMODULE += cc2538_rf
```

but I would only get

```
/usr/lib/gcc/arm-none-eabi/9.2.1/../../../arm-none-eabi/bin/ld: /home/benpicco/dev/RIOT/examples/gnrc_networking/bin/openmote-b/cc2538_rf/cc2538_rf.o: in function `cc2538_setup':
/home/benpicco/dev/RIOT/cpu/cc2538/radio/cc2538_rf.c:203: undefined reference to `cc2538_rf_driver'
collect2: error: ld returned 1 exit status
make: *** [/home/benpicco/dev/RIOT/examples/gnrc_networking/../../Makefile.include:587: /home/benpicco/dev/RIOT/examples/gnrc_networking/bin/openmote-b/gnrc_networking.elf] Error 1
```

is this expected?

benpicco commented on cpu/cc2538/radio/cc2538_rf_radio_ops.c:
```diff
@@ -0,0 +1,471 @@
+
+static uint8_t cc2538_min_be = 3;
+static uint8_t cc2538_max_be = 5;
+static int cc2538_retries = 4;
```
That's retransmissions, right? 
And wasn't the default 3? (Do we have a define for that?)

jia200x replied:
nope, this is the number of CSMA-CA retries (not the number of frame retransmissions). The standard defines Min_BE=3, Retries=4, Max_BE=5 for CSMA-CA params. Maybe this should be included in `net/ieee802154.h`

benpicco replied:
And better add a `cca` somewhere in the name to avoid that confusion.

jia200x replied:
done. I also added the parameters to `net/ieee802154.h`

benpicco commented on cpu/cc2538/radio/cc2538_rf_radio_ops.c:
```diff
@@ -0,0 +1,471 @@
+        /* Clear last program */
+        RFCORE_SFR_RFST = ISCLEAR;
+
+        /*  If the RSSI is not yet valid, busy loop */
+        RFCORE_SFR_RFST = SKIP_S_C |
+                          CC2538_CSP_SKIP_N_MASK |
+                          CC2538_CSP_SKIP_COND_RSSI;
+
+        RFCORE_SFR_RFST = LABEL;
+
+        RFCORE_SFR_RFST = RANDXY;
+        RFCORE_SFR_RFST = WAITX;
```
Where is the busy loop?

jia200x replied:
This is the internal Strobe Processor. This routine is written in the internal memory and executed with `RFCORE_SFR_RFST = ISSTART`. The busy loop is made with the SKIP_S instruction (23.14.8, Table 23-6 from Reference Manual)

jia200x replied:
should I mark this one as resolved too?

benpicco replied:
Would be good if you could expand on the comment.

benpicco commented on cpu/cc2538/radio/cc2538_rf_radio_ops.c:
```diff
@@ -0,0 +1,471 @@
+
+static int _len(ieee802154_dev_t *dev)
+{
+    (void) dev;
+   return rfcore_peek_rx_fifo(0) - IEEE802154_FCS_LEN;
```
```suggestion
   (void) dev;
   return rfcore_peek_rx_fifo(0) - IEEE802154_FCS_LEN;
```

jia200x replied:
done

benpicco commented on cpu/cc2538/radio/cc2538_rf_radio_ops.c:
```diff
@@ -0,0 +1,471 @@
+        RFCORE_SFR_RFST = DECZ;
+        RFCORE_SFR_RFST = INCMAXY | (cc2538_max_be & CC2538_CSP_INCMAXY_MAX_MASK);
+
+        RFCORE_SFR_RFST = RPT_C |
+                          CC2538_CSP_SKIP_N_MASK |
+                          CC2538_CSP_SKIP_COND_CSPZ;
```
```suggestion
        RFCORE_SFR_RFST = RPT_C
                        | CC2538_CSP_SKIP_N_MASK
                        | CC2538_CSP_SKIP_COND_CSPZ;
```
looks a bit nicer IMHO, but's just my personal taste.

jia200x replied:
done

benpicco commented on cpu/cc2538/radio/cc2538_rf_radio_ops.c:
```diff
@@ -0,0 +1,471 @@
+
+    }
+
+    if (flags_f1 & CSP_STOP) {
```
What does this mean?

jia200x replied:
The CSP_STOP interrupt is called when the routine from the Strobe Processor finishes. This is used to read the status of the CSMA-CA operation.

benpicco replied:
Better add that as a comment.

jia200x replied:
done!

benpicco commented on tests/ieee802154_hal/main.c:
```diff
@@ -0,0 +1,655 @@
+/*
+ * Copyright (C) 2015 Freie Universität Berlin
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     tests
+ * @{
+ *
+ * @file
+ * @brief       Test application for AT86RF2xx network device driver
+ *
+ * @author      Hauke Petersen <hauke.petersen@fu-berlin.de>
```
might need an update :wink:

jia200x replied:
maybe :see_no_evil:

jia200x replied:
done!

benpicco reviewed (commented):
Ah this is only the lowest layer.
Well I can confirm that the old netdev part still works with `examples/gnrc_networking`.

With the test I was wondering 'why is this implemented in the test?'

Is the plan to move this to the submac layer eventually?

Looking good btw!

jia200x commented:
> Is the plan to move this to the submac layer eventually?

Yes, exactly! Most of the stuff here are handled by the SubMAC :)

jia200x commented on cpu/cc2538/include/cc2538_rf.h:
```diff
@@ -68,6 +68,31 @@ extern "C" {
 
 #define CC2538_CRC_BIT_MASK         (0x80)
 
+#define CC2538_CCA_THR_MASK         (0x000000FF)
+
+#define CC2538_CCA_MODE_MASK        (0x18)
+#define CC2538_CCA_MODE_POS         (3U)
+
+#define CC2538_CSP_SKIP_INST_MASK   (0x70)
+#define CC2538_CSP_SKIP_INST_POS    (4U)
+
+#define CC2538_CSP_SKIP_N_MASK      (0x08)
+
+#define CC2538_CSP_SKIP_COND_CCA    (0x00)
+#define CC2538_CSP_SKIP_COND_CSPZ   (0x06)
+#define CC2538_CSP_SKIP_COND_RSSI   (0x07)
+
+#define CC2538_SFR_MTMSEL_MASK      (0x7)
+#define CC2538_SFR_MTMSEL_TIMER_P   (0x2)
+#define CC2538_MCTRL_SYNC_MASK      (0x2)
+#define CC2538_MCTRL_RUN_MASK       (0x1)
+
+#define CC2538_CSP_MCU_CTRL_MASK    (0x1)
+
+#define CC2538_CSP_INCMAXY_MAX_MASK (0x7)
+
+#define CC2538_RXENABLE_RXON_MASK   (0x80)
```
sure, I will

jia200x replied:
done!

jia200x commented on cpu/cc2538/radio/cc2538_rf.c:
```diff
@@ -28,7 +28,13 @@
+ *
+ * The period is set to the CSMA-CA Backoff Period Unit (20 symbols, 320 us)
+ */
+#define TIMER_PERIOD_LSB           (0xF2)
+#define TIMER_PERIOD_MSB           (0x29)
```
I will add a comment to clarify that

jia200x commented on cpu/cc2538/radio/cc2538_rf.c:
```diff
@@ -119,20 +128,35 @@ void cc2538_init(void)
+    RFCORE_SFR_MTM0 |= TIMER_PERIOD_LSB;
+    RFCORE_SFR_MTM1 |= TIMER_PERIOD_MSB;
+
+    RFCORE_SFR_MTMSEL &= ~CC2538_SFR_MTMSEL_MASK;
```
Yes, but because I had to set the 3 LSBs of RFCORE_SFR_MTMSEL to 001 (hence 146 and 148). For the next operation, I have to set those bits to 000 (so I have to clear it again).
If I do it the other way around the timer will start with the wrong configuration.

jia200x replied:
should I mark is as resolved?

benpicco replied:
Better add a comment.

jia200x commented on cpu/cc2538/radio/cc2538_rf_radio_ops.c:
```diff
@@ -0,0 +1,471 @@
+static int cc2538_retries = 4;
+
+static bool cc2538_cca_status;
+static bool cc2538_cca;
```
yes, I will add a comment.

jia200x replied:
ah, it actually means that the result of the CCA is due to a CCA request and not because of a failed CSMA-CA (just to distinguish which kind of confirmation should be generated). I already documented this.

jia200x replied:
done!

fjmolinas commented on cpu/cc2538/radio/cc2538_rf_radio_ops.c:
```diff
@@ -0,0 +1,471 @@
+{
+
+    (void) dev;
+    bool wait_sfd = RFCORE->XREG_FSMSTAT0bits.FSM_FFCTRL_STATE > 0x2 && RFCORE->XREG_FSMSTAT0bits.FSM_FFCTRL_STATE < 0x7;
```
Magic numbers? Can a define be used here?

jia200x replied:
oh yes! I totally forgot this one

jia200x replied:
done!

fjmolinas commented on cpu/cc2538/radio/cc2538_rf_radio_ops.c:
```diff
@@ -0,0 +1,471 @@
+            break;
+        case IEEE802154_RX_AACK_ENABLED:
+            RFCORE->XREG_FRMCTRL0bits.AUTOACK = 1;
+            RFCORE_XREG_FRMCTRL1 &= ~0x4;
```
Magic numbers? Can a define be used here?

jia200x replied:
here as well

jia200x replied:
done!

jia200x commented:
I think I addressed all comments

jia200x commented:
I also added the Kconfig symbols for the CSMA-CA params.

benpicco commented on cpu/cc2538/radio/cc2538_rf_radio_ops.c:
```diff
@@ -0,0 +1,475 @@
+        RFCORE_SFR_RFST = ISCLEAR;
+
+        /*  If the RSSI is not yet valid, busy loop */
+        RFCORE_SFR_RFST = SKIP_S_C |
+                          CC2538_CSP_SKIP_N_MASK |
+                          CC2538_CSP_SKIP_COND_RSSI;
```
```suggestion
        RFCORE_SFR_RFST = SKIP_S_C
                        | CC2538_CSP_SKIP_N_MASK
                        | CC2538_CSP_SKIP_COND_RSSI;
```

jia200x commented:
@benpicco done! :)

benpicco commented:
Please squash!

benpicco reviewed (approved):
I did some regression tests and the old `netdev` code path is still working.
So let's move on with this!

jia200x commented:
rebased and squashed!

benpicco commented:
Murdock is not happy

jia200x commented:
Fixed.

I also adapted to the changes from #14938 and removed references to other radios in the Makefile (only the cc2538_rf is supported so far, when the other radios are there we can update the Makefile)

jia200x commented:
thanks for the review!
